File: src/main/java/com/appzonegroup/creditclub/pos/util/TripleDesCipher.java
package com.appzonegroup.creditclub.pos.util;

import javax.crypto.Cipher;
import javax.crypto.SecretKey;
import javax.crypto.SecretKeyFactory;
import javax.crypto.spec.DESedeKeySpec;

/**
 * Created by mac on 1/29/19.
 */

public class TripleDesCipher {
    SecretKey key;

    public TripleDesCipher(byte [] rawkey) throws Exception
    {
        key = readKey(rawkey);
    }

    public  SecretKey readKey(byte[] rawkey) throws Exception
    {
        byte[] keyDes = new byte[24];
        System.arraycopy(rawkey, 0, keyDes, 0, 16);
        System.arraycopy(rawkey, 0, keyDes, 16, 8);
        DESedeKeySpec keyspec = new DESedeKeySpec(keyDes);
        SecretKeyFactory keyfactory = SecretKeyFactory.getInstance("DESede");
        return keyfactory.generateSecret(keyspec);
    }

    public byte[] encrypt(byte[] plain ) throws Exception
    {
        Cipher cipher = Cipher.getInstance("DESede/ECB/NoPadding");
        //final IvParameterSpec iv = new IvParameterSpec(new byte[8]);
        cipher.init(Cipher.ENCRYPT_MODE, key);
        byte[] encrypted = cipher.doFinal(plain);
        return encrypted;
    }
    public byte[] decrypt(byte[] cipher ) throws Exception
    {
        Cipher dcipher = Cipher.getInstance("DESede/ECB/NoPadding");
        //final IvParameterSpec iv = new IvParameterSpec(new byte[8]);
        dcipher.init(Cipher.DECRYPT_MODE, key);
        byte[] decrypted = dcipher.doFinal(cipher);
        return decrypted;
    }
}

File: poscore/src/main/java/com/creditclub/pos/utils/TripleDesCipher.kt
package com.creditclub.pos.utils

import java.security.InvalidKeyException
import java.security.NoSuchAlgorithmException
import java.security.spec.InvalidKeySpecException
import javax.crypto.*
import javax.crypto.spec.DESedeKeySpec

inline val ByteArray.asDesEdeKey: SecretKey
    @Throws(
        InvalidKeyException::class,
        NoSuchAlgorithmException::class,
        InvalidKeySpecException::class,
    )
    get() {
        val rawKey = this
        val keyDes = if (rawKey.size == 8) {
            val keyDes = ByteArray(16)
            System.arraycopy(rawKey, 0, keyDes, 0, 8)
            System.arraycopy(rawKey, 0, keyDes, 8, 8)

            keyDes
        } else {
            val keyDes = ByteArray(24)
            System.arraycopy(rawKey, 0, keyDes, 0, 16)
            System.arraycopy(rawKey, 0, keyDes, 16, 8)

            keyDes
        }

        val keySpec = DESedeKeySpec(keyDes)
        val keyFactory = SecretKeyFactory.getInstance("DESede")
        return keyFactory.generateSecret(keySpec)
    }

fun SecretKey.encrypt(data: ByteArray, transformation: String = "DESede/ECB/NoPadding"): ByteArray {
    val cipher = Cipher.getInstance(transformation)
    cipher.init(Cipher.ENCRYPT_MODE, this)
    return cipher.doFinal(data)
}

fun SecretKey.decrypt(data: ByteArray, transformation: String = "DESede/ECB/NoPadding"): ByteArray {
    val cipher = Cipher.getInstance(transformation)
    cipher.init(Cipher.DECRYPT_MODE, this)
    return cipher.doFinal(data)
}
